File: LeetCode/2382. Maximum Segment Sum After Removals/solution.cpp
class Solution {
public:
    vector<long long> maximumSegmentSum(vector<int>& nums, vector<int>& removeQueries) {
        int n = nums.size();
        // 求前缀和，方便我们直接求出区间的和
        vector<long long> f(n + 1);
        for (int i = 1; i <= n; i++) f[i] = f[i - 1] + nums[i - 1];

        // 记录已删除的下标
        set<int> st;
        // 放入下标 0 和 n + 1，这样就无需处理边界情况
        st.insert(0); st.insert(n + 1);
        // 记录所有子段和
        multiset<long long> ms;
        // 一开始只有一个子段和，也就是整个数组之和
        ms.insert(f[n]);
        vector<long long> ans;
        for (int pos : removeQueries) {
            pos++;
            // 找出第一个比 pos 大的已删除下标 R，以及最后一个比 pos 小的已删除下标 L
            auto it = st.upper_bound(pos);
            int L = *prev(it), R = *it;
            // 删除下标 pos，把它也放入 st 中
            st.insert(pos);
            // 删除子段 [L + 1, R - 1] 的和
            ms.erase(ms.find(f[R - 1] - f[L]));
            // 加入新子段 [L + 1, pos - 1] 的和
            if (pos - L - 1 > 0) ms.insert(f[pos - 1] - f[L]);
            // 加入新子段 [pos + 1, R - 1] 的和
            if (R - pos - 1 > 0) ms.insert(f[R - 1] - f[pos]);
            // 求出最大的子段和
            if (ms.empty()) ans.push_back(0);
            else ans.push_back(*prev(ms.end()));
        }
        return ans;
    }
};


// 作者：TsReaper
// 链接：https://leetcode.cn/circle/discuss/oZvJdG/view/sKjmHf/
// 来源：力扣（LeetCode）
// 著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。

File: LeetCode/2382. Maximum Segment Sum After Removals/solution.py
class Solution:
    def maximumSegmentSum(self, nums: List[int], removeQueries: List[int]) -> List[int]:
        n = len(nums)
        fa = list(range(n + 1))
        sum = [0] * (n + 1)
        def find(x: int) -> int:
            if fa[x] != x:
                fa[x] = find(fa[x])
            return fa[x]
        ans = [0] * n
        for i in range(n - 1, 0, -1):
            x = removeQueries[i]
            to = find(x + 1)
            fa[x] = to  # 合并 x 和 x+1
            sum[to] += sum[x] + nums[x]
            ans[i - 1] = max(ans[i], sum[to])
        return ans


# 作者：endlesscheng
# 链接：https://leetcode.cn/problems/maximum-segment-sum-after-removals/solution/by-endlesscheng-p61j/
# 来源：力扣（LeetCode）
# 著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。

class Solution:
    def maximumSegmentSum(self, nums: List[int], removeQueries: List[int]) -> List[int]:
        n = len(nums)
        parent = [i for i in range(n)]
        tot = [0]*n 
        ans = [0]*n 
        res = 0
        def find(cur):
            if parent[cur] != cur:
                parent[cur] = find(parent[cur])
            return parent[cur]
        def connect(u, v):
            ru, rv = find(u), find(v)
            if ru != rv:
                parent[ru] = rv 
                tot[rv] += tot[ru]
            return
        for i in range(n)[::-1]:
            q = removeQueries[i]
            tot[q] = nums[q]
            if q > 0 and tot[q-1] > 0:
                connect(q,q-1)
            if q+1 < n and tot[q+1] > 0:
                connect(q,q+1)
            if i > 0:
                res = max(res,tot[find(q)])
                ans[i-1] = res
        return ans 



from sortedcontainers import *
class Solution:
    def maximumSegmentSum(self, nums: List[int], removeQueries: List[int]) -> List[int]:
        n = len(nums)
        presums = [0]
        for x in nums:
            presums.append(presums[-1]+x)
        mp = [n]*n 
        intervals = SortedList([0])
        sums = SortedList([presums[-1]])
        ans = []
        for i in removeQueries:
            j = intervals.bisect_right(i)-1
            left = intervals[j]
            right = mp[left]
            intervals.remove(left)
            sums.remove(presums[right]-presums[left])
            if left < i:
                intervals.add(left)
                mp[left] = i 
                sums.add(presums[i]-presums[left])
            if i+1 < right:
                intervals.add(i+1)
                mp[i+1] = right
                sums.add(presums[right]-presums[i+1])
            if sums:
                ans.append(sums[-1])
            else:
                ans.append(0)
        return ans